### PR TITLE
convert Focus to TypeScript

### PR DESCRIPTION
**Pull request by unional**
This branch is based on `ts1` thus includes all changes in that branch.

This can be merged after #655 is merged.

I am opening this PR now to gather feedbacks.

### CODE REVIEW

**Comment by unional** [comment]
IMO the `addCommands` and `addMethod` should be avoided.
It is likely causing some bugs, especially the `focus` is a singleton.

File: src/api/focus/Focus.ts [hunk 1]
```diff
@@ -218,11 +232,11 @@ class Focus {
     return supported;
   }
 
-  isCommandSupported(cmd) {
+  isCommandSupported(cmd: string) {
     return this.supportedCommands.indexOf(cmd) !== -1;
   }
 
-  async _write_parts(parts, cb) {
+  async _write_parts(parts: any[], cb: any) {
```
Comment (unional) [inline src/api/focus/Focus.ts:243]: This method is not used anywhere, remove?

File: src/api/focus/Focus.ts [hunk 2]
```diff
@@ -285,38 +303,38 @@ class Focus {
...
-  addCommands(cmds) {
+  addCommands(cmds: CommandOverrides) {
     Object.assign(this.commands, cmds);
   }
 
-  addMethod(methodName, command) {
-    if (this[methodName]) {
-      const tmp = this[methodName];
-      this[methodName] = (...args) => {
+  addMethod(methodName: string, command: string) {
```
Comment (unional) [inline src/api/focus/Focus.ts:321]: This method while being used in two places,
the added methods are not being used.

I think this can be removed.

**Comment by unional** [comment]
Hi @alexpargon,

I notice that you migrated some code into `Device` and `Serial`, etc.

I take a look at the new code. IMO one concern about the new code is its testability.
The code that interacts with external (e.g. serial port) needs to be isolated and most business logic should be extract out of them as much as possible.

You can consider applying [Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html) or [Hexagonal Architecture](https://en.wikipedia.org/wiki/Hexagonal_architecture_(software))

**Comment by alexpargon** [comment]
Hey there @unional !

You are right about the new comms block, but it's a change that was required as the new methods of communication require further integration with the Node engine (HID), I have just implemented this feature as is but will better abstract it as soon as we release the Bluetooth capable Bazecor (HID), that will be my next priority because testing is a must and after your efforts, it's getting simpler to write good tests.

**Comment by unional** [comment]
Sounds good. Let me know if there is any area you need help to clean up / redesign / add tests.

I'm even thinking of rewriting kaleidoscope in Rust or Zig. 🤣

**Pull request merged**